### PR TITLE
Implement external arbiter mon

### PR DESCRIPTION
**Pull request by arttor**
Implements #14733, follow-up to discussion in #15210

First draft of ext mon implementaion. Ready for review. I was able to test it locally and it seems working for diffrent use-cases like adding/removing ext mon id to spec, upscaling/downcaling internal mons when ext mon is presented. 

Feature adds the following behaviour: If mon if is presented in `spec.mon.externalMonIDs: []` then rook will not manage it and **not** include such montiror in `monCount`. If external mon is also in quorum, then rook will add its ip address to all secrets and configMaps with mon endopoints. All other exising behaviour remains unchaged.

TODO:
 

- [x] add tests
- [x]  add docs

### CODE REVIEW

File: pkg/apis/ceph.rook.io/v1/types.go [hunk 1]
```diff
@@ -626,6 +626,10 @@ type MonSpec struct {
 	// +kubebuilder:pruning:PreserveUnknownFields
 	// +optional
 	VolumeClaimTemplate *VolumeClaimTemplate `json:"volumeClaimTemplate,omitempty"`
+	// ExternalMonIDs - optional list of monitor IDS which are deployed externallly and not managed by Rook.
+	// If set, Rook will not remove mons with given IDs from quorum.
+	// +optional
+	ExternalMonIDs []string `json:"externalMonIDs,omitempty"`
```
[the remark this answers is not in the document]
Comment (travisn) [reply pkg/apis/ceph.rook.io/v1/types.go:632]: For strech clusters, we also need to know which zone the arbiter belongs to. Perhaps we need to add this under stretch cluster config:
```yaml
  mon:
    stretchCluster:
      zones:
        - name: a
          arbiter: true
          external: <ID>
```

**Review by travisn — COMMENTED** [review]
Here are just a few high level comments. I still need to dig into more scenarios like mon failover, stretch cluster mon creation, etc. 

One question... With this feature, now `mon.count` would be 4 in stretch clusters, plus an external arbiter, right? There is a check somewhere that enforces 5 mons for a stretch cluster that likely still needs to be changed.

File: pkg/apis/ceph.rook.io/v1/types.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (arttor) [reply pkg/apis/ceph.rook.io/v1/types.go:632]: sorry, i forgot to mention it in documentation but my intention for this PR was exact opposite - use external mon as alternative to stretch cluster. 

I have not tested this particular use-case locally but if stretch cluster is enabled, then `externalMonIDs` should be ignored.

**Comment by arttor** [comment]
> 
> One question... With this feature, now `mon.count` would be 4 in stretch clusters, plus an external arbiter, right? There is a check somewhere that enforces 5 mons for a stretch cluster that likely still needs to be changed.

I was thinking to use it only for **non-stretched** cluster and current implementation follows this logic:
-  `mon.count` is only responsible for mons which are managed by Rook. If `mon.count` is different from number of mons running in local rook cluster, then rook will upscale/downscale local mons and ignore number of external monitors.
- if external mon is down rook will do nothing. the only thing rook does about external mon is: if mon id is in `externalMonIDs` rook will not remove it from quorum and add its ip to mon endpoints configmaps/secrets.

File: pkg/apis/ceph.rook.io/v1/types.go [hunk 2]
```diff
@@ -626,6 +626,10 @@ type MonSpec struct {
 	// +kubebuilder:pruning:PreserveUnknownFields
 	// +optional
 	VolumeClaimTemplate *VolumeClaimTemplate `json:"volumeClaimTemplate,omitempty"`
+	// ExternalMonIDs - optional list of monitor IDS which are deployed externallly and not managed by Rook.
```
[the remark this answers is not in the document]
Comment (arttor) [reply pkg/apis/ceph.rook.io/v1/types.go]: fixed

File: pkg/apis/ceph.rook.io/v1/types.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (travisn) [reply pkg/apis/ceph.rook.io/v1/types.go:632]: If this is only for the non-stretch cluster case, I am really curious at a deeper level how it is being used in the 2-datacenter scenario for which stretch cluster was specifically designed. It may mean extra risk or downtime or reliability issues when not using proper stretch clusters. While we can add this feature, it seems like a more interesting scenario to support the external arbiter for stretch clusters, but that could come in a future PR when needed.

File: pkg/daemon/ceph/client/info.go [hunk 3]
```diff
@@ -36,13 +36,14 @@ import (
...
-	CephCred      CephCred
-	Monitors      map[string]*MonInfo
-	CephVersion   cephver.CephVersion
-	Namespace     string
-	OwnerInfo     *k8sutil.OwnerInfo
+	FSID             string
+	MonitorSecret    string
+	CephCred         CephCred
+	InternalMonitors map[string]*MonInfo
+	ExtArbiterMons   map[string]*MonInfo
```
Comment (travisn) [inline pkg/daemon/ceph/client/info.go:43]: Let's reserve the term "arbiter" only for stretch clusters. Since this feature is not for stretch clusters, let's just call them external mons.
```suggestion
	ExternalMons   map[string]*MonInfo
```

File: pkg/operator/ceph/cluster/mon/health.go [hunk 4]
```diff
@@ -377,20 +383,98 @@ func (c *Cluster) checkHealth(ctx context.Context) error {
 	return nil
 }
 
+func (c *Cluster) reconcileExtArbiterMons(ctx context.Context, quorumStatus cephclient.MonStatusResponse) (cephclient.MonStatusResponse, error) {
```
Comment (travisn) [inline pkg/operator/ceph/cluster/mon/health.go:386]:
```suggestion
func (c *Cluster) reconcileExternalMons(ctx context.Context, quorumStatus cephclient.MonStatusResponse) (cephclient.MonStatusResponse, error) {
```

File: pkg/daemon/ceph/client/info.go (hunk 3 above)
Comment (arttor) [reply pkg/daemon/ceph/client/info.go:43]: sure, will remove arbiter naming from all places

**Review by travisn — COMMENTED** [review]
LGTM, just a few nits

**Comment by arttor** [comment]
PR is ready for review. Added tests and docs and will do another round of local tests just in case.

**Review by travisn — APPROVED** [review]

**Comment by arttor** [comment]
Updated docs, tested edge-cases locally. Ready for review/merge.

File: Documentation/Storage-Configuration/Advanced/ceph-mon-health.md [hunk 5]
```diff
@@ -124,3 +124,43 @@ CephCluster CR:
...
 - `spec.network.Provider` : When updated from being empty to "host", Rook fails over all monitors, configuring them to enable or disable host networking.
 - `spec.network.multiClusterService`: When enabled or disabled, Rook fails over all monitors, configuring them to start (or stop) using service IPs compatible with the multi-cluster service.
+
+## External Monitors
+
+!!! attention
+    This feature is experimental.
+
+It is possible to have both Rook-managed and external monitors in the same Rook cluster.
+One use case for this is 2 datacenter aka 2-AZ (Availability Zone) setup. For 2-AZ setup, Zone outage will lead to loss of k8s control plane and a half of worker nodes hosting Rook mons.
```
Comment (travisn) [inline Documentation/Storage-Configuration/Advanced/ceph-mon-health.md:134]:
```suggestion
One use case for this is 2 datacenter aka 2-AZ (Availability Zone) setup. For 2-AZ setup, Zone outage will lead to loss of the k8s control plane and half of the worker nodes hosting Rook mons.
```

File: Documentation/Storage-Configuration/Advanced/ceph-mon-health.md [hunk 6]
```diff
@@ -124,3 +124,43 @@ CephCluster CR:
...
+In this case remaining half of Rook cluster will not be able to form quorum and will be in a stuck state even if half of worker nodes are still up.
```
Comment (travisn) [inline Documentation/Storage-Configuration/Advanced/ceph-mon-health.md:135]:
```suggestion
In this case remaining half of the Rook cluster will not be able to form quorum and will be in a stuck state even if the other half of worker nodes are still up.
```

File: Documentation/Storage-Configuration/Advanced/ceph-mon-health.md [hunk 7]
```diff
@@ -124,3 +124,43 @@ CephCluster CR:
...
+To avoid this situation, external mons can be used to form quorum and keep the cluster running.
+
+If you have external monitors that are not managed by Rook, you have to inform Rook about them, otherwise Rook will remove unknown mons from the quorum. This is done by setting the `mon.externalMonIDs` field in the CephCluster CR. Internal `mon.count` ignores a number of `mon.externalMonIDs`. In other words, if `mon.count = 2`, Rook will spawn 2 internal mons no matter how many external mons are in the cluster and what is their health state. External monitors are supported only for local Rook cluster running in normal mode, meaning that `mon.externalMonIDs` will be ignored for external cluster (`spec.external.enabled: true`) or for `stretchedCluster`.
```
Comment (travisn) [inline Documentation/Storage-Configuration/Advanced/ceph-mon-health.md:138]: In the docs it is preferred to avoid "you", and rather speak in third person.
```suggestion
If there are external monitors, Rook must be aware of them, otherwise Rook will remove the unknown mons from the quorum. This is done by setting the `mon.externalMonIDs` field in the CephCluster CR. The `mon.count` ignores the number of `mon.externalMonIDs`. For example, if `mon.count = 2`, Rook will create two internal mons no matter how many external mons are in the cluster and no matter what their health state might be. External monitors are supported only for the local Rook cluster running in normal mode. The external mons will be ignored for external clusters and stretch clusters.
```

File: Documentation/Storage-Configuration/Advanced/ceph-mon-health.md [hunk 8]
```diff
@@ -124,3 +124,43 @@ CephCluster CR:
...
+      mon:
+        # Spawn 2 Mons - one Mon in each AZ managed by Rook
+        count: 2
+        allowMultiplePerNode: false
+        # ID of external Mon
+        externalMonIDs:
+        - ext-mon-1 
+    ```
+
+    This will tell Rook to run 2 internal monitors in the cluster and to keep the external monitor with the ID `ext-mon-1` if it will be presented in the quorum.
```
Comment (travisn) [inline Documentation/Storage-Configuration/Advanced/ceph-mon-health.md:155]:
```suggestion
    This will tell Rook to create two internal monitors in the cluster and to keep the external monitor with the ID `ext-mon-1` if it is found in the quorum.
```

File: Documentation/Storage-Configuration/Advanced/ceph-mon-health.md [hunk 9]
```diff
@@ -124,3 +124,43 @@ CephCluster CR:
...
+    It is also possible to add `externalMonIDs` to existing Cluster.
```
Comment (travisn) [inline Documentation/Storage-Configuration/Advanced/ceph-mon-health.md:156]:
```suggestion
    It is also possible to add `externalMonIDs` to an existing Cluster.
```

File: Documentation/Storage-Configuration/Advanced/ceph-mon-health.md [hunk 10]
```diff
@@ -124,3 +124,43 @@ CephCluster CR:
...
+2. Wait until the Rook cluster and internal mons are up and running.
+3. Manually deploy external monitor with the ID `ext-mon-1` somewhere outside of the Rook cluster and its availability zones. See [ceph guide](https://docs.ceph.com/en/reef/rados/operations/add-or-rm-mons/#adding-removing-monitors) on deploying monitors.
```
Comment (travisn) [inline Documentation/Storage-Configuration/Advanced/ceph-mon-health.md:158]:
```suggestion
3. Manually deploy an external monitor with the ID `ext-mon-1` outside of the Rook cluster and its availability zones. See [ceph guide](https://docs.ceph.com/en/latest/rados/operations/add-or-rm-mons/#adding-removing-monitors) on deploying monitors.
```

File: Documentation/Storage-Configuration/Advanced/ceph-mon-health.md [hunk 11]
```diff
@@ -124,3 +124,43 @@ CephCluster CR:
...
+4. Move external Mon to [disallow-mode](https://docs.ceph.com/en/reef/rados/operations/change-mon-elections/#rados-operations-disallow-mode) to make sure that it won't be elected as a leader. The purpose of external mode is maintaining quorum to elect a new leader in case of zone outage. External Mon will likely have higher latency to the cluster so it is not desired to be elected.
```
Comment (travisn) [inline Documentation/Storage-Configuration/Advanced/ceph-mon-health.md:159]:
```suggestion
4. Move the external Mon to [disallow-mode](https://docs.ceph.com/en/reef/rados/operations/change-mon-elections/#rados-operations-disallow-mode) to make sure that it won't be elected as a leader. The purpose of external mode is maintaining quorum to elect a new leader in case of a zone outage. The external Mon will likely have higher latency to the cluster so it should not be elected.
```

File: Documentation/Storage-Configuration/Advanced/ceph-mon-health.md [hunk 12]
```diff
@@ -124,3 +124,43 @@ CephCluster CR:
...
+5. Check that the external monitor is in the quorum by running `ceph status` or `ceph quorum_status` from the toolbox.
+6. Check that external mon is added to Rook mon endpoints:
```
Comment (travisn) [inline Documentation/Storage-Configuration/Advanced/ceph-mon-health.md:161]:
```suggestion
6. Check that the external mon is added to the Rook mon endpoints:
```

**Review by travisn — CHANGES_REQUESTED** [review]
Some suggestions on the doc, thanks!

**Comment by arttor** [comment]
applied doc suggestions and squashed commits.

File: PendingReleaseNotes.md [hunk 13]
```diff
@@ -13,3 +13,5 @@ Object:
     See https://github.com/rook/rook/pull/15376 for more information.
 
 ## Features
+
+- Support external mons for local Rook cluster (see [#14733](https://github.com/rook/rook/issues/14733)).
```
Comment (arttor) [inline PendingReleaseNotes.md:17]: @travisn added pending release notes

**Review by travisn — APPROVED** [review]
